projection/ibc_channel/view/ibc_channels.go

```diff
@@ -430,10 +430,222 @@ func (ibcChannelsView *IBCChannels) List(
+			}
+			return nil, nil, fmt.Errorf("error scanning channel row: %v: %w", err, rdb.ErrQuery)
+		}
+
+		createdAtBlockTime, parseErr := createdAtTimeReader.Parse()
+		if parseErr != nil {
+			return nil, nil, fmt.Errorf("error parsing createdAtBlockTime: %v: %w", parseErr, rdb.ErrQuery)
+		}
+		channel.CreatedAtBlockTime = *createdAtBlockTime
+
+		lastActivityBlockTime, parseErr := lastActivityTimeReader.Parse()
+		if parseErr != nil {
+			return nil, nil, fmt.Errorf("error parsing lastActivityBlockTime: %v: %w", parseErr, rdb.ErrQuery)
+		}
+		channel.LastActivityBlockTime = *lastActivityBlockTime
+
+		if err = jsoniter.UnmarshalFromString(bondedTokensJSON, &channel.BondedTokens); err != nil {
+			return nil, nil, fmt.Errorf("error unmarshalling channel bonded_tokens JSON: %v: %w", err, rdb.ErrQuery)
+		}
+
+		channels = append(channels, channel)
+	}
+
+	paginationResult, err := rDbPagination.Result()
+	if err != nil {
+		return nil, nil, fmt.Errorf("error preparing pagination result: %v", err)
+	}
+
+	return &channels, paginationResult, nil
+}
+
+func (ibcChannelsView *IBCChannels) ListChannelsGroupByChainId(
+	order IBCChannelsListOrder,
+	filter IBCChannelsListFilter,
+	pagination *pagination.Pagination,
+) (
+	*[]ChainChannels,
+	*pagination.PaginationResult,
+	error,
+) {
+	// First to get all chainId
+	stmtBuilder := ibcChannelsView.rdb.StmtBuilder.Select(
+		"counterparty_chain_id",
+	).From(
+		"view_ibc_channels",
+	).Distinct()
+
+	rDbPagination := rdb.NewRDbPaginationBuilder(
+		pagination,
+		ibcChannelsView.rdb,
+	).BuildStmt(stmtBuilder)
+	sql, sqlArgs, err := rDbPagination.ToStmtBuilder().ToSql()
+	if err != nil {
+		return nil, nil, fmt.Errorf("error building select distinct counterparty_chain_id SQL: %v, %w", err, rdb.ErrBuildSQLStmt)
+	}
+
+	rowsResult, err := ibcChannelsView.rdb.Query(sql, sqlArgs...)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error executing select distinct counterparty_chain_id SQL: %v: %w", err, rdb.ErrQuery)
+	}
+	defer rowsResult.Close()
+
+	chainIDs := make([]string, 0)
+	for rowsResult.Next() {
+		var chainID string
+		if err = rowsResult.Scan(
+			&chainID,
+		); err != nil {
+			if errors.Is(err, rdb.ErrNoRows) {
+				return nil, nil, rdb.ErrNoRows
+			}
+			return nil, nil, fmt.Errorf("error scanning channel row counterparty_chain_id: %v: %w", err, rdb.ErrQuery)
+		}
+
+		chainIDs = append(chainIDs, chainID)
+	}
+
+	// Then, loop through chainId, get each chain's channel list
+	chainChannelsList := make([]ChainChannels, 0)
+	for _, chainID := range chainIDs {
+
+		var chainChannels ChainChannels
+
+		chainChannels.ChainID = chainID
+		channels, _, listChannelsErr := ibcChannelsView.listIBCChannelsByChainID(order, filter, pagination, chainID)
+		if listChannelsErr != nil {
+			return nil, nil, fmt.Errorf("error listing channel of a specific chain: %v", err)
+		}
+		chainChannels.Channels = *channels
+
+		chainChannelsList = append(chainChannelsList, chainChannels)
+	}
```

Comment:
This will be a performance issue of 1+N query if there are many chain ids, why not do a single query and group them programmatically on scan?

Reply:
Thanks for the advice!

This implementation is trying to return all channels under a specific chain. But yes, it will cause performance issue... and hard to do pagination for the chain's channel list.

I will update the code according to your comment!